core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.collections;
+
+import com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+
+/**
+ * A cache specifically designed to contain locks and will NOT evict any entries
+ * that are in use.
+ *
+ * @param <K> key to the cache
+ * @param <V> often a lock type
+ */
+public class LockCache<K, V> {
+  /**
+   * Node containing value and other information to be stored in the cache.
+   *
+   * @param <V> value type
+   */
+  public class ValNode<V> {
```

Comment:
What do you think of returning a `LockResource` from `get`, instead of the `ValNode` type? Then we don't need to expose the reference counting outside of `LockCache`.

Reply:
yeah that seems safer, so we don't accidentally forget to decRef. 
If we are making it specific to locks, maybe we don't need the generic class either?

Reply:
Actually it seems more complicated than that, some methods currently needs direct access to the locks rather than having the ability to lock.